### PR TITLE
[IS-1177] add recovery mode

### PR DESCRIPTION
**Pull request by yakkle**
## Added recovery mode to loopchain to recover from unexpected failure.

- replace GetStatus grpc call with REST api `/status/peer`
  - `GetStatus` grpc call is deprecated
- added `Recovery` state in `channel_statemachine`
- wait until there are more than `2f+2` nodes in the `Recovery` state
- add as many blocks as `RELEASE_RECOVERY_BLOCK_COUNT`, and then release recovery mode.
- enable RECOVERY_MODE by shell environment variable (values : `True|true|False|false`)
  ```shell
  $ RECOVERY_MODE=true loop -d -o loopchain_conf.json
  ```
- fixed to get `highest_block_height` of recovery
- fixed to prevent transition to `BlockSync` when `Recovery` state
- If the new recovery mode node has a block height difference of greater than
`RELEASE_RECOVERY_BLOCK_COUNT(5)` blocks, release the recovery mode.

### CODE REVIEW

File: loopchain/blockchain/blockchain.py [hunk 1]
```diff
@@ -964,10 +972,14 @@ def confirm_prev_block(self, current_block: Block):
...
                 if self.last_block.header.hash == current_block.header.prev_hash:
                     logging.warning(f"Already added block hash({current_block.header.prev_hash.hex()})")
+                    self.last_unconfirmed_block = current_block
+                    # TODO: NEED REVIEW - check this logic should be limited by below condition
+                    """
                     if ((current_block.header.complained and self.__block_manager.epoch.complained_result)
                             or self.last_block.header.prep_changed):
                         utils.logger.debug("reset last_unconfirmed_block by complain block or first block of new term.")
                         self.last_unconfirmed_block = current_block
+                    """
```
Comment (yakkle) [inline loopchain/blockchain/blockchain.py:982]: @icon-project/consensus 
please review this condition is necessary.

File: loopchain/peer/peer_outer_service.py [hunk 2]
```diff
@@ -161,8 +162,11 @@ def AnnounceUnconfirmedBlock(self, request, context):
         except AttributeError:
             round_ = 0
 
+        utils.logger.warning(f"request.from_recovery : {request.from_recovery!r}")
+        from_recovery = request.from_recovery if request.from_recovery else False
+
         asyncio.run_coroutine_threadsafe(
-            channel_stub.async_task().announce_unconfirmed_block(request.block, round_),
+            channel_stub.async_task().announce_unconfirmed_block(request.block, round_, from_recovery),
```
Comment (yakkle) [inline loopchain/peer/peer_outer_service.py:169]: restore this changes.

File: loopchain/channel/channel_statemachine.py [hunk 3]
```diff
@@ -151,6 +169,9 @@ def _consensus_on_enter(self, *args, **kwargs):
     def _blockheightsync_on_enter(self, *args, **kwargs):
         self.evaluate_network()
 
+    def _recovery_mode_on_enter(self, *args, **kwargs):
+        pass
+
```
Comment (yakkle) [inline loopchain/channel/channel_statemachine.py:174]: remove this lines.

**Review by windies21 — APPROVED** [review]

File: loopchain/peer/peer_outer_service.py (hunk 2 above)
Comment (yakkle) [reply loopchain/peer/peer_outer_service.py:169]: not restore. mistake.

File: loopchain/peer/block_manager.py [hunk 4]
```diff
@@ -140,12 +142,28 @@ def _send_unconfirmed_block(self, block_: Block, target_reps_hash, round_: int):
...
         block_dumped = self.blockchain.block_dumps(block_)
-        ObjectManager().channel_service.broadcast_scheduler.schedule_broadcast(
+        send_kwargs = {"block": block_dumped, "round_": round_, "channel": self.__channel_name}
+
+        release_recovery_mode = False
+        if conf.RECOVERY_MODE:
+            from loopchain.tools.recovery import Recovery
+            if self.blockchain.block_height <= Recovery.release_block_height():
+                util.logger.info(f"broadcast block({block_.header.height}) from recovery node")
+                send_kwargs.update({"from_recovery": True})
```
Comment (goldworm-icon) [inline loopchain/peer/block_manager.py:152]:
```suggestion
                send_kwargs["from_recovery"] = True
```

File: loopchain/peer/block_manager.py [hunk 5]
```diff
@@ -140,12 +142,28 @@ def _send_unconfirmed_block(self, block_: Block, target_reps_hash, round_: int):
...
+        self.__channel_service.broadcast_scheduler.schedule_broadcast(
             "AnnounceUnconfirmedBlock",
-            loopchain_pb2.BlockSend(block=block_dumped, round_=round_, channel=self.__channel_name),
+            loopchain_pb2.BlockSend(**send_kwargs),
             reps_hash=target_reps_hash
         )
 
+        if release_recovery_mode:
+            conf.RECOVERY_MODE = False
+            util.logger.info(f"recovery mode released at {self.blockchain.block_height}")
```
Comment (goldworm-icon) [inline loopchain/peer/block_manager.py:165]: Why don't you move this code block to line 155?

File: loopchain/peer/block_manager.py [hunk 6]
```diff
@@ -140,12 +142,28 @@ def _send_unconfirmed_block(self, block_: Block, target_reps_hash, round_: int):
...
+        send_kwargs = {"block": block_dumped, "round_": round_, "channel": self.__channel_name}
+
+        release_recovery_mode = False
+        if conf.RECOVERY_MODE:
+            from loopchain.tools.recovery import Recovery
+            if self.blockchain.block_height <= Recovery.release_block_height():
+                util.logger.info(f"broadcast block({block_.header.height}) from recovery node")
+                send_kwargs.update({"from_recovery": True})
+
+            if self.blockchain.block_height >= Recovery.release_block_height():
```
Comment (goldworm-icon) [inline loopchain/peer/block_manager.py:154]: Is that right to set True to release_recovery_mode at Recovery.release_block_height()?

File: loopchain/tools/recovery.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    def set_target_list(self, target_list: List[str]) -> None:
+        regex = re.compile(r":([0-9]{2,5})$")
+        for target in target_list:
+            port = regex.search(target).group(1)
+            new_port = f"{int(port) + conf.PORT_DIFF_REST_SERVICE_CONTAINER}"
+            endpoint = target.replace(port, new_port)
+            self._endpoints.append(endpoint)
+
+        fault: int = int((len(self._endpoints) - 1) / 3)
```
Comment (goldworm-icon) [inline loopchain/tools/recovery.py:30]:
```suggestion
        fault: int = (len(self._endpoints) - 1) // 3
```

File: loopchain/tools/recovery.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                                     f"my_block_height({self._last_block_height}), "
+                                     f"highest_block_height({main_highest_block_height})")
+                        conf.RECOVERY_MODE = False
+                        return
+
+            logging.debug(f"highest_block_height: {Recovery._highest_block_height}")
+            logging.info(f"recovery_mode quorum : {recovery_quorum}")
+            if recovery_quorum > self._min_quorum:
+                if waiting_recovery > 1:
+                    await asyncio.sleep(conf.RECOVERY_CHECK_INTERVAL + 1)
```
[the remark this answers is not in the document]
Comment (yakkle) [reply loopchain/tools/recovery.py]: removed.

File: loopchain/peer/block_manager.py (hunk 5 above)
Comment (yakkle) [reply loopchain/peer/block_manager.py:165]: IMO, `RECOVERY_MODE` should be set disabled after generator broadcast unconfirmed block.

File: loopchain/peer/block_manager.py (hunk 6 above)
Comment (yakkle) [reply loopchain/peer/block_manager.py:154]: Condition is correct.

File: loopchain/blockchain/blockchain.py [hunk 9]
```diff
@@ -608,11 +608,20 @@ def __add_block(self, block: Block, confirm_info, need_to_write_tx_info=True, ne
...
                 f"CHANNEL : {self.__channel_name}")
             utils.logger.debug(f"ADDED BLOCK HEADER : {block.header}")
 
+            if conf.RECOVERY_MODE:
+                from loopchain.tools.recovery import Recovery
+                utils.logger.debug(f"release recovery_mode block height : {Recovery.release_block_height()}")
+                if (channel_service.state_machine.state in ('Vote', 'BlockSync')
+                        and block.header.height >= Recovery.release_block_height()):
+                    conf.RECOVERY_MODE = False
+                    logging.info(f"recovery mode released at {block.header.height}")
```
[the remark this answers is not in the document]
Comment (yakkle) [reply loopchain/blockchain/blockchain.py:623]: you can find that on [line 619](https://github.com/icon-project/loopchain/pull/583/files#diff-6e4d4360ac9e07410adaff6572f27638066f7605ad419011471345a15053d019R619).
```py
utils.logger.debug(f"release recovery_mode block height : {Recovery.release_block_height()}")
```

**Review by goldworm-icon — APPROVED** [review]

**Pull request merged**